### PR TITLE
Feature/remote generic substitution

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

Add remote type support for generic substitution. This enables us to remove schemagen code from wasm. 

## Testing
New tests added

### For dApp Developers

The non-fungible resource creation interface has been updated. See the new manifest syntax within `transaction/examples`.

### For Internal Integrators

Various substate breaking changes, notably
- `NonFungibleDataSchema`
- `KeyValueStoreDataSchema`
- `BlueprintDefinitionInit`
- `BlueprintInterface`

### CODE REVIEW

File: radix-engine-common/src/types/type_identifier.rs [hunk 1]
```diff
@@ -54,7 +54,17 @@ impl ScopedTypeId {
...
+#[derive(Debug, Clone, PartialEq, Eq, Hash, ManifestSbor, ScryptoSbor)]
 pub enum GenericSubstitution {
     Local(ScopedTypeId),
+    /// Currently supports default version of blueprint only.
+    /// New variants can be added for specific version of blueprint in the future.
+    Remote(BlueprintTypeId),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, ManifestSbor, ScryptoSbor)]
+pub struct BlueprintTypeId {
```
Comment (dhedey) [inline radix-engine-common/src/types/type_identifier.rs:66]: Suggestion: `BlueprintTypeName`?

Comment (dhedey) [reply radix-engine-common/src/types/type_identifier.rs:66]: (Relatedly, perhaps `EventTypeIdentifier` should be `EmittedEventName` :hmm:)

File: radix-engine-interface/src/api/key_value_store_api.rs [hunk 2]
```diff
@@ -1,59 +1,84 @@
...
+/// Less flexible than previous revision, as mixed type origin is not allowed, but
+/// better for client-side optimization
 #[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor, ManifestSbor)]
-pub struct KeyValueStoreGenericArgs {
-    pub additional_schema: Option<VersionedScryptoSchema>,
-    pub key_type: GenericSubstitution,
-    pub value_type: GenericSubstitution,
-    pub allow_ownership: bool,
+pub enum KeyValueStoreDataSchema {
+    // TODO: ignore this variant in Scrypto for smaller code size
```
Comment (dhedey) [inline radix-engine-interface/src/api/key_value_store_api.rs:16]: Nice thought 👍 . VersionedScryptoSchema is a large type.

File: radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs [hunk 3]
```diff
@@ -304,30 +290,61 @@ pub struct NonFungibleResourceManagerMintSingleRuidTypedInput<T> {
...
 
 impl NonFungibleData for () {
     const MUTABLE_FIELDS: &'static [&'static str] = &[];
 }
 
 impl NonFungibleDataSchema {
-    pub fn new_schema<N: NonFungibleData>() -> Self {
+    /// Arguments:
+    /// * [`package_address`] - The package address to use for resolving types defined within this package
+    pub fn new_local<N: NonFungibleData>(package_address: PackageAddress) -> Self {
```
Comment (dhedey) [inline radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs:313]: Maybe `new_local_with_self_package_replacement` ?

File: radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs [hunk 4]
```diff
@@ -304,30 +290,61 @@ pub struct NonFungibleResourceManagerMintSingleRuidTypedInput<T> {
...
 }
 
 impl NonFungibleData for () {
     const MUTABLE_FIELDS: &'static [&'static str] = &[];
 }
 
 impl NonFungibleDataSchema {
-    pub fn new_schema<N: NonFungibleData>() -> Self {
+    /// Arguments:
+    /// * [`package_address`] - The package address to use for resolving types defined within this package
```
Comment (dhedey) [inline radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs:312]: Is it worth saying that `None` are replaced with the current package?

File: radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs [hunk 5]
```diff
@@ -304,30 +290,61 @@ pub struct NonFungibleResourceManagerMintSingleRuidTypedInput<T> {
...
+    pub fn new_local<N: NonFungibleData>(package_address: PackageAddress) -> Self {
         let mut aggregator = TypeAggregator::<ScryptoCustomTypeKind>::new();
-        let non_fungible_type = aggregator.add_child_type_and_descendents::<N>();
+        let type_id = aggregator.add_child_type_and_descendents::<N>();
+        let mut schema = generate_full_schema(aggregator);
+        replace_self_package_address(&mut schema, package_address);
```
Comment (dhedey) [inline radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs:317]: Can you double-check that we validate that the replacement has been done when validating the schema?

File: radix-engine/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs [hunk 6]
```diff
@@ -491,13 +492,60 @@ impl NonFungibleResourceManagerBlueprint {
...
+            NonFungibleDataSchema::Remote {
+                type_id,
+                mutable_fields,
+            } => {
+                let (schema, scoped_type_id) = api.resolve_blueprint_type(&type_id)?;
+                let mutable_indices =
+                    Self::validate_non_fungible_schema(&schema, scoped_type_id.1, mutable_fields)?;
+                Ok((
+                    GenericArgs {
+                        additional_schema: Some(schema),
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs]: Done

File: radix-engine/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs [hunk 7]
```diff
@@ -491,13 +492,60 @@ impl NonFungibleResourceManagerBlueprint {
...
-        non_fungible_schema: &NonFungibleDataSchema,
+        schema: &VersionedScryptoSchema,
+        local_type_id: LocalTypeId,
+        mutable_fields: &IndexSet<String>,
     ) -> Result<IndexMap<String, usize>, RuntimeError> {
         let mut mutable_field_index = indexmap!();
 
         // Validate schema
-        validate_schema(non_fungible_schema.schema.v1()).map_err(|e| {
+        validate_schema(schema.v1()).map_err(|e| {
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs]: Done

File: radix-engine/src/system/system_type_checker.rs [hunk 8]
```diff
@@ -197,6 +203,18 @@ where
 
                         (schema, type_id.1, SchemaOrigin::Instance)
                     }
+                    GenericSubstitution::Remote(type_id) => {
+                        let (schema, scoped_type_id) = self.get_blueprint_type_schema(&type_id)?;
+
+                        (
+                            schema,
+                            scoped_type_id.1,
+                            SchemaOrigin::Blueprint(BlueprintId {
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/system/system_type_checker.rs]: Done

File: radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs (hunk 5 above)
Comment (dhedey) [reply radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs:317]: Hmm I've checked in `custom_schema.rs` - looks like we don't do this validation

File: radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs (hunk 4 above)
Comment (iamyulong) [reply radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs:312]: Done

File: radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs (hunk 5 above)
Comment (dhedey) [reply radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs:317]: But looks like this is resolved automatically to the blueprint in `matches_with_origin` anyway

Comment (dhedey) [reply radix-engine-interface/src/blueprints/resource/non_fungible/non_fungible_resource_manager.rs:319]: Ignore all this thread, as per discussions

**Review by talekhinezh — APPROVED** [review]

**Pull request merged**